### ISSUE TITLE
Percent encoding a query parameter whose value contains the equal sign '='

### ISSUE THREAD

**Issue by dbrakman** [gruns/furl#99]
I want to create this query string:
```python
str(f.query) == 'where=1%3D1'
```

Expecting query parameters and values to be decoded automatically, I tried this:
```python
f = furl.furl()
f.args['where'] = '1=1'
```
but got an undesired result:
```python
str(f.query) == 'where=1=1'
```

Alternatively, I tried encoding the parameter value before passing it to the furl object:
```python
f = furl.furl()
f.args['where'] = urllib.quote_plus('1=1')
```
but found the percent encoding encoded again:
```python
str(f.query) == 'where=1%253D1'
```

How can I create the literal query string `where=1%3D1`?

**Comment by gruns** [comment]
Let's back up a bit.

  1. Is it a hard requirement that the query be `where=1%3D1` and not
    `where=1=1`? If so, why?

  2. Does a library or piece of software (e.g. a web server) incorrectly parse
     `where=1=1`? If so, which one, and how does it parse `which=1=1`?

Unfortunately the exact `key=value` format within a query string isn't
standardized. In turn, the format differs between library and software
implementations. RFC 3986 only specifies which characters are valid query
characters, not the `key=value` format commonly used to encode keys and values
within said query.

Zooming out a bit, the equal sign `=` is a perfectly valid query character and
is permitted, un-encoded, in the query. In other words this:

```python
>>> from furl import furl
>>> furl('https://www.google.com/').set(query='===3===').url
'https://www.google.com/?===3==='
```

is perfectly valid. While `===3===` doesn't adhere to the `a=b&c=d` de facto
`key=value` specification, it's a perfectly valid query as per RFC 3986.

So, should furl allow this query as-is? Or should furl percent encode it,
perhaps against the user's will, to `https://www.google.com/?%3D%3D%3D3%3D%3D%3D`?

I think a more measured approach by furl would be to detect explicit `key=value`
pairs and encode the `=` signs in such pairs, but otherwise leave the query
as-is. That would accomodate both `?===3===` and `which=1%3D1`.

**Comment by dbrakman** [comment]
To answer your questions:

1. To my knowledge, it's not a hard requirement anywhere. It's just one of the possible query parameter strings, and I'm curious to know how to create it using furl.

2. I could have sworn a proprietary dependency was incorrectly parsing `where=1=1`, but, upon the further investigation this prompted, apparently not.

I think I see your point about valid characters in RFC 3986 and why encoding queries involves a nontrivial design decision. I had wondered if there was or should be a flag for indicating a literal query string, but I can't say what's best for furl.

In the potential approach you described, would we get `where=1%3D1` from something like the code below?
```python
f = furl.furl()
f.set(args='1%3D1')
```

**Comment by gruns** [comment]
Thank you for answering the aforementioned questions. That's quite helpful.

The first compromising approach that comes to mind is to percent encode `=`s in
values if and only if the associated key is empty. Otherwise, leave any `=`s in
the value as-is.

furl stores queries like `?===3===` as `omdict1D([('', '==3===')])`. `key=value`
query params are stored like `omdict1D([('key', 'value')])`. In turn, it's
straightforward for furl to differentiate between queries like `===3===` and
`which=1%3D1`.

I'll ruminate on this more.

furl should adopt the least surprising approach.

**Comment by gruns** [comment]
Fixed in https://github.com/gruns/furl/commit/e7de07d0700d9bd8f4cfb7f4c985812657af17fe, which will ship in furl v1.1 shortly.

**Comment by gruns** [comment]
Fixed in furl v1.1.

```python
>>> from furl import furl
>>> f = furl()
>>> f.args['where'] = '1=1'
>>> f.url
'?where=1%3D1'
```

Upgrade furl with

```
pip install furl --upgrade
```

Thank you, again, for bringing this issue to my attention @dbrakman. Don't
hesitate to let me know if there's anything else I can do for you.
